### PR TITLE
Distinguish LFS object errors to ignore missing objects during migration

### LINKED ISSUE #31137: Migrating repo with LFS causes 500 error
### Description

https://gitea.com/Shuenhoy/vtk-m is migrated from https://gitlab.kitware.com/vtk/vtk-m. 
Accessing the `data` folder will trigger 500 error.

Whether selecting `migrate LFS files` is the same.
In the demo site, https://demo.gitea.com/Shuenhoy/vtk-m will not trigger 500 error, but it doesn't treat the files as LFS files at all. 

### Gitea Version

1.22.0

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

image

(On my instance, there is an extra message)
```
An error occurred:

getFileReader, LFS Meta object does not exist
```

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Docker

### Database

None

### PR DESCRIPTION
**Pull request by wolfogre**
Fix #31137.

Replace #31623 #31697.

When migrating LFS objects, if there's any object that failed (like some objects are losted, which is not really critical), Gitea will stop migrating LFS immediately but treat the migration as successful.

This PR checks the error according to the [LFS api doc](https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md#successful-responses).

> LFS object error codes should match HTTP status codes where possible:
> 
> - 404 - The object does not exist on the server.
> - 409 - The specified hash algorithm disagrees with the server's acceptable options.
> - 410 - The object was removed by the owner.
> - 422 - Validation error.

If the error is `404`, it's safe to ignore it and continue migration. Otherwise, stop the migration and mark it as failed to ensure data integrity of LFS objects.

And maybe we should also ignore others errors (maybe `410`? I'm not sure what's the difference between "does not exist" and "removed by the owner".), we can add it later when some users report that they have failed to migrate LFS because of an error which should be ignored.

### CODE REVIEW

File: modules/lfs/shared.go [hunk 1]
```diff
@@ -64,6 +66,39 @@ type ObjectError struct {
...
+)
+
+func (e *ObjectError) Error() string {
+	return fmt.Sprintf("[%d] %s", e.Code, e.Message)
+}
+
+func (e *ObjectError) Unwrap() error {
+	switch e.Code {
+	case 404:
+		return ErrObjectNotFound
```
Comment (delvh) [inline modules/lfs/shared.go]: Can't we use the existing `ErrNotExist`?
I don't want to maintain a separate `NotExist` error, that'll only create problems

Comment (delvh) [reply modules/lfs/shared.go]: Additionally, we probably want to catch `401`/`403` as `ErrPermissionDenied`

Comment (wolfogre) [reply modules/lfs/shared.go]: > Can't we use the existing `ErrNotExist`?

I prefer to alias the existing `ErrNotExist`, see 7ebf151258e14af260136722195fe0fe9059ad71

So `errors.Is(objectError, lfs.ErrObjectNotExist)` and `errors.Is(objectError, util.ErrNotExist)` are both fine.

> Additionally, we probably want to catch `401`/`403` as `ErrPermissionDenied`

I don't think so, please read the [doc](https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md?#successful-responses). There's no `401`/`403` for "object error", if there's are some permission errors, they will be "response errors" and we cannot get the objects (or the object errors).

And we don't need to specifically handle the permission error; just let the migration fail like handling other errors.

**Review by delvh — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by wolfogre** [comment]
Please hold on. I need to hear @6543's opinions.

**Comment by wolfogre** [comment]
I'll block it for another two days and merge it if there's no against.

**Comment by wxiaoguang** [comment]
> Please hold on. I need to hear @6543's opinions.

Actually I would never expect some maintainers to respond


image

**Comment by 6543** [comment]
but looking at the diff in a glance it looks way better :)

**Pull request merged**